src/lib/mina_base/parties_logic.ml

```diff
@@ -0,0 +1,292 @@
+             ; global_state: 'global_state
+             ; .. > )
+           t
+    | Set_account_if :
+        'bool * 'ledger * 'account * 'inclusion_proof
+        -> ( 'ledger
+           , < bool: 'bool
+             ; ledger: 'ledger
+             ; inclusion_proof: 'inclusion_proof
+             ; account: 'account
+             ; .. > )
+           t
+    | Modify_global_excess :
+        'global_state * ('amount -> 'amount)
+        -> ( 'global_state
+           , < global_state: 'global_state ; amount: 'amount ; .. > )
+           t
+    | Modify_global_ledger :
+        'global_state * ('ledger -> 'ledger)
+        -> ( 'global_state
+           , < global_state: 'global_state ; ledger: 'ledger ; .. > )
+           t
+    | Party_token_id :
+        'party
+        -> ('token_id, < party: 'party ; token_id: 'token_id ; .. >) t
+    | Check_auth_and_update_account :
+        [`Is_first | `Is_not_first] * 'party * 'account * 'all_parties * 'ip
+        -> ( 'account * 'bool
+           , < inclusion_proof: 'ip
+             ; bool: 'bool
+             ; party: 'party
+             ; all_parties: 'all_parties
+             ; account: 'account
+             ; .. > )
+           t
+    | Balance :
+        'account
+        -> ('amount, < account: 'account ; amount: 'amount ; .. >) t
+    | Finalize_local_state :
+        'bool * 'local_state
+        -> (unit, < local_state: 'local_state ; bool: 'bool ; .. >) t
+end
+
+type 'e handler = {perform: 'r. ('r, 'e) Eff.t -> 'r}
+
+module type Inputs_intf = sig
+  module Bool : Bool_intf
+
+  module Ledger : Ledger_intf with type bool := Bool.t
+
+  module Account : Iffable with type bool := Bool.t
+
+  module Amount : Amount_intf with type bool := Bool.t
+
+  module Token_id : Token_id_intf with type bool := Bool.t
+
+  module Parties : Parties_intf with type bool := Bool.t
+end
+
+module Make (Inputs : Inputs_intf) = struct
+  open Inputs
+
+  let apply (type global_state)
+      (step_or_start : [`Start of Inputs.Parties.t | `Step])
+      (h :
+        (< global_state: global_state ; amount: Amount.t ; bool: Bool.t ; .. >
+         as
+         'env)
+        handler)
+      ((global_state : global_state), (local_state : _ Local_state.t)) =
+    let open Inputs in
+    let (party, remaining), local_state =
+      match step_or_start with
+      | `Step ->
+          (Inputs.Parties.pop local_state.parties, local_state)
+      | `Start parties ->
+          let prev_finished = Inputs.Parties.is_empty local_state.parties in
+          Bool.assert_ prev_finished ;
+          let init_party, parties = Inputs.Parties.pop parties in
+          ( (init_party, parties)
+          , (* The initiator party is omitted from the all_parties list. *)
+            {local_state with all_parties= parties; token_id= Token_id.default}
+          )
+    in
+    let local_state = {local_state with parties= remaining} in
+    let a, inclusion_proof =
+      (* Also checks the inclusion proof. *)
```

Comment:
Should this comment be deleted? It seems like the inclusion check is done by `Check_inclusion` below.